File: backend/Application/Mappers/EducationMapper.cs
using Application.DTOs.Responses;
using Application.Interfaces;
using Domain.Entities;

namespace Application.Mappers;

public class EducationMapper : IEducationMapper
{
    private readonly ITechnologyMapper _technologyMapper;
    private readonly ISkillMapper _skillMapper;
    private readonly IToolingMapper _toolingMapper;

    public EducationMapper(
        ITechnologyMapper technologyMapper,
        ISkillMapper skillMapper,
        IToolingMapper toolingMapper)
    {
        _technologyMapper = technologyMapper;
        _skillMapper = skillMapper;
        _toolingMapper = toolingMapper;
    }

    public EducationResponse Map(Education education)
    {
        return new EducationResponse
        {
            Id = education.Id,
            Program = education.Program,
            StartYear = education.StartYear,
            EndYear = education.EndYear,

            Organization = new OrganizationResponse
            {
                Id = education.Organization.Id,
                Name = education.Organization.Name,
                Type = education.Organization.Type.ToString(),
                WebsiteUrl = education.Organization.WebsiteUrl,

                Media = education.Organization.Media is null
                    ? null
                    : new MediaResponse
                    {
                        Path = education.Organization.Media.Path,
                        AltText = education.Organization.Media.AltText,
                        Type = education.Organization.Media.Type
                    }
            },

            Credentials = education.Credentials
                .Select(x => new CredentialResponse
                {
                    Id = x.Id,
                    Name = x.Name,
                    Type = x.Type.ToString(),
                    Status = x.Status.ToString(),
                    Year = x.Year,
                    CredentialUrl = x.CredentialUrl,
                    DocumentUrl = x.DocumentUrl
                })
                .ToList(),

            Sections = education.Sections
                .OrderBy(x => x.SortOrder)
                .Select(section => new EducationSectionResponse
                {
                    Id = section.Id,
                    Title = section.Title,

                    Technologies = section.Technologies
                        .OrderBy(x => x.Name)
                        .Select(_technologyMapper.Map)
                        .ToList(),

                    Skills = section.Skills
                        .OrderBy(x => x.Name)
                        .Select(_skillMapper.Map)
                        .ToList(),

                    Tooling = section.Tooling
                        .OrderBy(x => x.Name)
                        .Select(_toolingMapper.Map)
                        .ToList()
                })
                .ToList()
        };
    }
}

File: frontend/src/mappers/educationMapper.js
export function mapEducation(data) {
    if (!data) {
        throw new Error("Education data is missing");
    }

    return {
        id: data.id,
        program: data.program ?? "",
        startYear: data.startYear,
        endYear: data.endYear,

        organization: data.organization
            ? {
                id: data.organization.id,
                name: data.organization.name ?? "",
                type: data.organization.type ?? "",
                websiteUrl: data.organization.websiteUrl ?? null,

                media: data.organization.media
                    ? {
                        path: data.organization.media.path ?? "",
                        altText: data.organization.media.altText ?? "",
                        type: data.organization.media.type,
                    }
                    : null,
            }
            : null,

        credentials: (data.credentials ?? []).map((credential) => ({
            id: credential.id,
            name: credential.name ?? "",
            type: credential.type ?? "",
            status: credential.status ?? "",
            year: credential.year ?? null,
            credentialUrl: credential.credentialUrl ?? null,
            documentUrl: credential.documentUrl ?? null,
        })),

        sections: (data.sections ?? []).map((section) => ({
            id: section.id,
            title: section.title ?? null,

            technologies: (section.technologies ?? []).map(
                (technology) => ({
                    id: technology.id,
                    name: technology.name ?? "",
                    usage: technology.usage ?? "",

                    media: technology.media
                        ? {
                            path: technology.media.path ?? "",
                            altText: technology.media.altText ?? "",
                            type: technology.media.type,
                        }
                        : null,
                })
            ),

            skills: (section.skills ?? []).map((skill) => ({
                id: skill.id,
                name: skill.name ?? "",
            })),

            tooling: (section.tooling ?? []).map((tool) => ({
                id: tool.id,
                name: tool.name ?? "",

                media: tool.media
                    ? {
                        path: tool.media.path ?? "",
                        altText: tool.media.altText ?? "",
                        type: tool.media.type,
                    }
                    : null,
            })),
        })),
    };
}